src/legacy/core_plugins/kibana/public/dashboard/migrations/migrate_to_730_panels.ts

```diff
@@ -0,0 +1,304 @@
+import { i18n } from '@kbn/i18n';
+import { DEFAULT_PANEL_WIDTH, DEFAULT_PANEL_HEIGHT } from '../dashboard_constants';
+import { PanelUtils } from '../panel/panel_utils';
+import {
+  RawSavedDashboardPanelTo60,
+  RawSavedDashboardPanel630,
+  RawSavedDashboardPanel640To720,
+  RawSavedDashboardPanel730ToLatest,
+  RawSavedDashboardPanel610,
+  RawSavedDashboardPanel620,
+} from './types';
+import {
+  SavedDashboardPanelTo60,
+  SavedDashboardPanel620,
+  SavedDashboardPanel630,
+  SavedDashboardPanel610,
+} from '../types';
+
+const PANEL_HEIGHT_SCALE_FACTOR = 5;
+const PANEL_HEIGHT_SCALE_FACTOR_WITH_MARGINS = 4;
+const PANEL_WIDTH_SCALE_FACTOR = 4;
+
+/**
+ * Note!
+ *
+ * The 7.3.0 migrations reference versions that are quite old because for a long time all of this
+ * migration logic was done ad hoc in the code itself, not on the indexed data (migrations didn't even
+ * exist at the point most of that logic was put in place).
+ *
+ * So you could have a dashboard in 7.2.0 that was created in 6.3 and it will have data of a different
+ * shape than some other dashboards that were created more recently.
+ *
+ * Moving forward migrations should be simpler since all 7.3.0+ dashboards should finally have the
+ * same data.
+ */
+
+function isPre61Panel(
+  panel: unknown | RawSavedDashboardPanelTo60
+): panel is RawSavedDashboardPanelTo60 {
+  return (panel as RawSavedDashboardPanelTo60).row !== undefined;
+}
+
+function is61Panel(panel: unknown | RawSavedDashboardPanel610): panel is RawSavedDashboardPanel610 {
+  const panelVersion = PanelUtils.parseVersion((panel as RawSavedDashboardPanel610).version);
+  return panelVersion && panelVersion.major === 6 && panelVersion.minor === 1;
+}
+
+function is62Panel(panel: unknown | RawSavedDashboardPanel620): panel is RawSavedDashboardPanel620 {
+  const panelVersion = PanelUtils.parseVersion((panel as RawSavedDashboardPanel620).version);
+  return panelVersion && panelVersion.major === 6 && panelVersion.minor === 2;
+}
+
+function is63Panel(panel: unknown | RawSavedDashboardPanel630): panel is RawSavedDashboardPanel630 {
+  const panelVersion = PanelUtils.parseVersion((panel as RawSavedDashboardPanel630).version);
+  return panelVersion && panelVersion.major === 6 && panelVersion.minor === 3;
+}
+
+function is640To720Panel(
+  panel: unknown | RawSavedDashboardPanel640To720
+): panel is RawSavedDashboardPanel640To720 {
+  const panelVersion = PanelUtils.parseVersion((panel as RawSavedDashboardPanel640To720).version);
+  return (
+    panelVersion &&
+    ((panelVersion.major === 6 && panelVersion.minor > 3) ||
+      (panelVersion.major === 7 && panelVersion.minor < 3))
+  );
+}
+
+// Migrations required for 6.0 and prior:
+// 1. (6.1) migrate size_x/y/row/col into gridData
+// 2. (6.2) migrate uiState into embeddableConfig
+// 3. (6.3) scale grid dimensions
+// 4. (6.4) remove columns, sort properties
+// 5. (7.3) make sure panelIndex is a string
+function migratePre61PanelToLatest(
```

Comment:
Since we have a `migrate610PanelToLatest` later on, should this function just migrate 6.0 and prior to 6.1, and then let the `migrate610PanelToLatest` move it from 6.1 to latest? Same for the remaining functions in this file. Most of them appear to have some redundancy. It seems as if it might be easier to do this similar to the way Kibana migrations do them, where each function just moves its arguments forward one version, and you pipe the data through all the functions to move it up to the latest shape.

That might dry things up a bit. Might over-complicate it, too.

Reply:
I thought about doing that, but ultimately decided against it.  The one part that was tripping me up was this:

```
      w: panel.size_x ? panel.size_x * PANEL_WIDTH_SCALE_FACTOR : DEFAULT_PANEL_WIDTH,
      h: panel.size_y ? panel.size_y * heightScaleFactor : DEFAULT_PANEL_HEIGHT,
```

If you want to just do a step from 6.1 to 6.2, you have to pick not the current default panel width but the default panel width before we actually bumped up the row and columns.  Maybe that's not a good enough reason but it started me in this direction, and mimicked the original ad-hoc/in place migration code (which also was 6.1 -> latest, etc), so I stuck with it.